include/boost/graph/link_cut_trees.hpp

```diff
@@ -0,0 +1,277 @@
+        template <class Element>
+        void splay(Element x)
+        {
+            for (Element x_parent = get_parent(x); x != x_parent; x_parent = get_parent(x))
+            {
+                const Element x_grandparent = get_parent(x_parent);
+                const ElementChildMap &x_side = get_side(x);
+                const ElementChildMap &x_parent_side = get_side(x_parent);
+
+                if (x_grandparent == x_parent)
+                    rotate(x_parent, x_side);
+                else if (&x_side == &x_parent_side)
+                {
+                    rotate(x_grandparent, x_parent_side);
+                    rotate(x_parent, x_side);
+                }
+                else
+                {
+                    rotate(x_parent, x_side);
+                    rotate(x_grandparent, x_parent_side);
+                }
+            }
+        }
+
+        template <class Element>
+        Element expose(Element x)
+        {
+            Element r = x;
+            while (true)
+            {
+                const Element x_successor = get_successor(find_path(x));
+                const std::pair<Element, Element> uv = split(x);
+                if (x != uv.first)
+                    put_successor(uv.first, x);
+                r = join(r, x, uv.second);
+                if (x == x_successor)
+                    break;
+                x = x_successor;
+            }
+            put_successor(r, r);
+            return r;
+        }
+
+        template <class Element>
+        Element find_path(Element x)
+        {
+            splay(x);
+            return x;
+        }
+
+        template <class Element>
+        Element find_tail(Element x)
+        {
+            while (get(right, x) != x)
+                x = get(right, x);
+            splay(x);
+            return x;
+        }
+
+        template <class Element>
+        Element join(Element u, Element v, Element w)
+        {
+            if (u != v)
+                put(parent, u, v);
+            if (w != v)
+                put(parent, w, v);
+            put(left, v, u);
+            put(right, v, w);
+            return v;
+        }
+
+        template <class Element>
+        std::pair<Element, Element> split(Element x)
+        {
+            splay(x);
+            Element x_left = get(left, x);
+            Element x_right = get(right, x);
+            if (x_left != x)
+                put(parent, x_left, x_left);
+            if (x_right != x)
+                put(parent, x_right, x_right);
+            put(left, x, x);
+            put(right, x, x);
+            return std::make_pair<>(x_left, x_right);
+        }
+    };
+    
+    
+    template <class ID = identity_property_map, 
+        class InverseID = boost::unordered_map<typename property_traits<ID>::value_type, typename property_traits<ID>::key_type>, 
+        class IndexMapContainer = boost::unordered_map<typename property_traits<ID>::value_type, typename property_traits<ID>::value_type> >
```

Comment:
Boost.Unordered has a new, much faster, hash map called `unordered_flat_map`: https://www.boost.org/doc/libs/develop/libs/unordered/doc/html/unordered.html

It's not compliant with `unordered_map`, but I presume we don't need the extra features of the standard interface and can drop it in here.